File: Hello LeetCode/LeetCodeBook-Java/src/com/algorithm/链表/_203_移除链表元素.java
package com.algorithm.链表;

/**
 * https://leetcode-cn.com/problems/remove-linked-list-elements/
 */
public class _203_移除链表元素 {
    public ListNode removeElements(ListNode head, int val) {
        ListNode newHead = new ListNode(0);
        newHead.next = head;
        ListNode current = newHead;

        while (current.next != null) {
            if (current.next.val == val) {
                current.next = current.next.next;
            } else {
                current = current.next;
            }
        }

        return newHead.next;
    }

    //自己写的
    public ListNode removeElements1(ListNode head, int val) {
        ListNode newHead = new ListNode(0);
        newHead.next = head;
        ListNode node = head;

        ListNode prev = newHead;

        while (node != null) {
            if (node.val == val) {
                prev.next = node.next;
                node = node.next;
            } else {
                prev = node;
                node = node.next;
            }
        }

        return newHead.next;
    }
}

File: Hello LeetCode/LeetCodeBook-Swift/Leetcode-Book/Leetcode-Book/LinkedList/_203_移除链表元素.swift
//
//  _203_移除链表元素.swift
//  Leetcode-Book
//
//  Created by gaoliang on 2022/1/20.
//

import Foundation

class _203_移除链表元素_Solution {
    
    func removeElements(_ head: ListNode?, _ val: Int) -> ListNode? {
        let dummy = ListNode(val - 1, head)
        var node: ListNode? = dummy
        
        while node?.next != nil {
            if node?.next!.val == val {
                node?.next = node?.next?.next
            } else {
                node = node?.next
            }
        }
        
        return dummy.next
    }
}

func test_203_移除链表元素_Solution() {
//    let head = switchNumbersToLists(nums: [1,2,6,3,4,5,6])
    let head = switchNumbersToLists(nums: [7,7,7,7])
    
    print(_203_移除链表元素_Solution().removeElements(head, 7))
}
